File: src/main/java/ShaderLoader.java
import org.lwjgl.opengl.Display;
import org.lwjgl.util.glu.GLU;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;

import static org.lwjgl.opengl.GL11.GL_FALSE;
import static org.lwjgl.opengl.GL11.GL_NO_ERROR;
import static org.lwjgl.opengl.GL11.glGetError;
import static org.lwjgl.opengl.GL20.*;
import static org.lwjgl.opengl.GL20.glGetProgramInfoLog;
import static org.lwjgl.opengl.GL20.glValidateProgram;

public class ShaderLoader {

  private static int createShader(String filename, int type) {
    StringBuilder shaderSource = new StringBuilder();
    int shaderID = 0;

    File file = new File(filename);

    if(!file.exists()) throw new IllegalArgumentException("File not found: " + file.getAbsolutePath());

    try {
      BufferedReader reader = new BufferedReader(new FileReader(file));
      String line;
      while ((line = reader.readLine()) != null) {
        shaderSource.append(line).append("\n");
      }
      reader.close();
    } catch (IOException e) {
      System.err.println("Could not read file: " + e.getMessage());
      e.printStackTrace();
      System.exit(-1);
    }

    shaderID = glCreateShader(type);
    glShaderSource(shaderID, shaderSource);
    glCompileShader(shaderID);

    if (glGetShaderi(shaderID, GL_COMPILE_STATUS) == GL_FALSE) {
      int maxLength = glGetShaderi(shaderID, GL_INFO_LOG_LENGTH);
      String info = glGetShaderInfoLog(shaderID, maxLength);
      throw new IllegalArgumentException("Could not compile " + file.getAbsolutePath() + ": " + info);
    }

    exitOnGLError("createShader");

    return shaderID;
  }

  public static int initializeProgram(String vertexShaderFilename, String fragmentShaderFilename) {
    // Load the vertex shader
    int vsId = createShader(vertexShaderFilename, GL_VERTEX_SHADER);
    // Load the fragment shader
    int fsId = createShader(fragmentShaderFilename, GL_FRAGMENT_SHADER);

    // Create a new shader program that links both shaders
    int pId = glCreateProgram();

    glAttachShader(pId, vsId);
    exitOnGLError("attachVertexShader");

    glAttachShader(pId, fsId);
    exitOnGLError("attachFragmentShader");;

    glLinkProgram(pId);
    if(glGetProgrami(pId, GL_LINK_STATUS) == GL_FALSE)
    {
      int infoLogLength = glGetProgrami(pId, GL_INFO_LOG_LENGTH);
      String infoLog = glGetProgramInfoLog(pId, infoLogLength);
      throw new IllegalArgumentException("Linker failure: " + infoLog);
    }

    exitOnGLError("linkProgram");

    glValidateProgram(pId);

    exitOnGLError("initializeProgram");
    return pId;
  }

  private static void exitOnGLError(String errorMessage) {
    int errorValue = glGetError();

    if (errorValue != GL_NO_ERROR) {
      String errorString = GLU.gluErrorString(errorValue);
      System.err.println("ERROR - " + errorMessage + ": " + errorString);

      if (Display.isCreated()) Display.destroy();
      System.exit(-1);
    }
  }
}

File: src/main/scala/opengl/ShaderLoader.scala
package opengl

import java.io.{File, IOException}
import java.util.Map

import org.lwjgl.opengl.GL11._
import org.lwjgl.opengl.GL20._
import org.lwjgl.opengl.GL32._
import org.lwjgl.util.glu.GLU

import scala.collection.JavaConverters._
import scala.io.Source


object ShaderLoader {
  def createShader(filename: String, shaderType: Int): Int = {
    var shaderID = 0
    val file = new File(filename)
    if (!file.exists) throw new IllegalArgumentException("File not found: " + file.getAbsolutePath)
    if (!file.canRead) throw new IllegalArgumentException("Cannot read file: " + file.getAbsolutePath)
    val shaderSource = try {
      Source.fromFile(file).mkString
    }
    catch {
      case e: IOException => {
        throw new IllegalArgumentException("Could not read file: " + e.getMessage)
      }
    }
    shaderID = glCreateShader(shaderType)
    glShaderSource(shaderID, shaderSource)
    glCompileShader(shaderID)
    if (glGetShaderi(shaderID, GL_COMPILE_STATUS) == GL_FALSE) {
      val maxLength = glGetShaderi(shaderID, GL_INFO_LOG_LENGTH)
      val info = glGetShaderInfoLog(shaderID, maxLength)
      throw new IllegalArgumentException("Could not compile " + file.getAbsolutePath + ": " + info)
    }
    checkForGlError("createShader")
    shaderID
  }

  def initializeProgram(shaderTypeToId: Map[Int, Int]): Int = {
    val programId = glCreateProgram

    shaderTypeToId.asScala map { case (key, value) =>
      glAttachShader(programId, value)
      checkForGlError("attachShader:" + shaderTypeToString(key))
    }

    glLinkProgram(programId)

    if (glGetProgrami(programId, GL_LINK_STATUS) == GL_FALSE) {
      val infoLogLength = glGetProgrami(programId, GL_INFO_LOG_LENGTH)
      val infoLog = glGetProgramInfoLog(programId, infoLogLength)
      throw new IllegalArgumentException("Linker failure: " + infoLog)
    }
    shaderTypeToId.values.asScala map { shaderId =>
      glDetachShader(programId, shaderId)
    }
    checkForGlError("linkProgram")
    glValidateProgram(programId)
    checkForGlError("initializeProgram")
    programId
  }

  private def shaderTypeToString(shaderType: Int) = shaderType match {
    case GL_VERTEX_SHADER => "vertex"
    case GL_GEOMETRY_SHADER => "geometry"
    case GL_FRAGMENT_SHADER => "fragment"
    case _ => throw new IllegalArgumentException("Invalid shader type: " + shaderType)
  }

  private def checkForGlError(errorMessage: String) {
    val errorValue = glGetError
    if (errorValue != GL_NO_ERROR) {
      val errorString = GLU.gluErrorString(errorValue)
      throw new IllegalStateException("GL Error - " + errorMessage + ": " + errorString)
    }
  }
}

